Allow external program arguments and expand initial ~

mreppen opened the pull request:
Allows external program devices command to pass arguments (not working since 2016-11-19).

Also expands initial ~ and ~user.

mreppen commented:
I should mention: In `call_prog_with_args` I didn't add expansion of tilde for Windows, but could do so.

However, I don't understand the extra manipulation of the Windows command in `call_prog_with_args` (to me the manipulation seems unnecessary compared to `callprogram`), which is why I didn't want to change it.

MAKOMO commented:
Max, thanks for your work. Could you give a quick example where the external program command breaks in the original version. Is this an issue only on Windows or also on the other two platforms (Linux/MacOS X). I am asking as those external calls seems to be very platform dependent and I have received recently several patches to solve specific problems and don't want to break things for those again.

MAKOMO commented:
There is currently a discussion on HB on a related issue on Linux. 

[http://www.home-barista.com/home-roasting/artisan-on-linux-text-to-speech-t45786.html#p519293](http://www.home-barista.com/home-roasting/artisan-on-linux-text-to-speech-t45786.html#p519293)

However, I cannot duplicate this on my setups. For me things work here as expected. So things might also be very (system) configuration dependent. Wonder how we could make this more robust in general.

mreppen commented:
`echo "100,100"` gives `Exception: callprogram(): [Errno 2] No such file or directory @line 30262` under Linux. This is what I fix by splitting the string (using shlex). Alternatively, setting `shell=True` should also solve it, but that I haven't actively tested (although already used in `def call_prog_with_args`). https://docs.python.org/2/library/subprocess.html#frequently-used-arguments

Let me explain what else is going on in the PR:
- Run `os.path.expanduser` to expand tildes on the strings to be executed.
- Reordered the if statements splitting based on platform (they were logically sound, but confused me). This only changes the actual call for the case `platf == "Windows and not (sys.version < '3')` where the change is to run with `shell=True`. If the previously mentioned bug is also present under Windows, I expect this to fix it. Otherwise, according to the documentation, this does essentially not change anything under Windows.

mreppen commented:
As I mentioned, `call_prog_with_args` has several lines I don't understand the purpose of (and also use `shell=` differently). Assuming there are no issues with `callprogram`, since the execution of commands should be the same, maybe it would be worthwhile to rewrite the former to work as the latter. This would possibly reduce the number of special cases and as a result also robustness.

That is not a part of this PR, but I could create another one or add it, if you are interested.
If something comes out of the home-barista discussion, I could have a look at that as well.

MAKOMO commented:
Dear Max,

I am now on it. Your changes look good. However, I am still struggling understanding why _echo "100,100" gives Exception: callprogram(): [Errno 2] No such file or directory @line 30262_. I tested such a script on my Mac and that works as expected (using the code from v1.0). 

EDIT: now I get the issue. If you paste that _echo "100,100"_ directly in the external program path as a command instead of a path, the original code fails. Right, but why would one want to do this!?

What is nice on your code is that it expands the users home directory. Works on my Mac. However, your code has an issue exactly spaces in the path (maybe I am missing something). 

With "~/Desktop/Fo lder/test2.sh" (without or with the quotes) pasted as external program path I receive "No such file or directory @line 30313" with your code as shlex.split(cmd_str) results in ['/Users/luther/Desktop/Fo', 'lder/test2.sh']. Without the shlex.split() everything runs fine. Hm. Supporting file paths with spaces seems to me more important than supporting shell commands directly, or am I wrong!?

The discussion on HB revealed that there was a space placed wrong in the script that made the script fail to run. So no issue detected there.

mreppen commented:
I agree that spaces in paths are important. However, that already doesn't work for me both with the latest commit 396f3988421383fa9b044402205110597ed11132, nor 47922d0eab5b7b09bf64daa9fa2d8753b18fe2c0 where my branch diverged.

In fact, spaces is broken (for me) **without** `shlex.split`. What causes your exception is that `os.path.expanduser` does not expand the tilde, since it is preceded by `"`.

I have updated the pull request with the fix.

MAKOMO commented:
Excellent Max! This works for me on Mac OS X now as expected. Now what to do with this pull request (after I committed all changes already)? Should I still click "Merge pull request" or just "Close and comment" (I am not a Git/GitHub expert). Unrelated: are you working on Linux? would you be interested in supporting me to move the Linux builds to Python3.5/latestQt/PyQt on that platform? one would mostly need to move to pyinstaller replacing this outdated bbfreeze along the lines of the RPi build script.

mreppen commented:
I have never **received** a PR, but think "Merge pull request" is the right thing to do. At least merging is the way to go when working in Git directly (w/o Github).

Yes, I'm running on Linux and would be happy to help.
I would have to read up on Qt, artisan code, etc, but think it'll be all right.

mreppen commented:
@MAKOMO I'm sorry I misread your question; my eyes skipped one line (the full sentence) after "expected."

In this case, closing would have been enough, since it was already committed (which is what I missed). Otherwise, "merge" would automatically add the changes, so that you wouldn't need to commit them separately.
